Keyname overhaul (#3447)
ElektraInitiative/libelektra: doc/keynames/README.md

> # Key Names in Elektra

Author: @markus2330 @mpranj I wasn't quite sure where to put this (and the accompanying Python file), so I made a new directory for now. If you have better ideas, let me know...

## Preface

> This document is a full explanation of how Key Names work in Elektra.

Reviewer: Why the capitalization? As we write "key-value", maybe we want "key-name"?

I did not read it now (I saw that there are still TODOs). Please tell me when it is ready to review.

Author: > Why the capitalization? As we write "key-value", maybe we want "key-name"?

The capitalization was very much intentional. It is a common sign of a proper noun in English. I used it to signify that e.g. "Key Name" means something very specific and is not just the name of a key. I actually capitalized all the terms a defined and I think I also italicized them in their definitions.

It also makes regex-replaces easy, if defined terms are capitalized, so I can replace it with "key-name" if you want to. Although in most of codebase we use "keyname", AFAIK. But my spell-checker really didn't like that (my hands don't like typing keyname either, so many typos....) so I avoided that.

> I did not read it now (I saw that there are still TODOs). Please tell me when it is ready to review.

As stated in the other comment, it is now ready for review.

Reviewer: > It is a common sign of a proper noun in English.

Yes, proper nouns are capitalized but technical terms never get proper nouns (with the rare exceptions of one-of-a-kind items like the Internet). It is a very common thing to (re)define technical terms but this does not make them proper nouns.

In Elektra we define technical terms in the glossary doc/help/elektra-glossary.md and use them everywhere in this meaning but not capitalized.

Please also do so for key names (or key-names or keynames if you want, as long as it is everywhere the same). Otherwise it only brings inconsistency without any benefit.

>  spell-checker, typos

You can add your common typos in scripts/sed

- A Key Name Part: `version`
- A Part Separator: `/`
- A Key Name Part: `info`
- The invisible null terminator

> _Note:_ It might seem strange, that there is a part **separator** before the first part.
> This makes sense, because then the Part Separator always introduces a new part.
> A better fitting description would be "part introducer".
> But since we commonly call `/` a separator, we will stick to this terminology.

For Keys in the Namespace "Cascading", we omit both the Namespace itself and as well as the Namespace separator:

```
/elektra/version/info
```

We also have a special rule for the Root Keys, i.e. the Key Names with zero Key Name Parts and just a Namespace.
According to above rules `system:` would be the Escaped Name of the Root Key for the Namespace "System" and the Escaped Name for the Root Key of the Namespace "Cascading" would be an empty string.
But this is not the case.
We use `system:/` and `/` instead.
The exact details will be explored later, but for now just remember, that an (Escaped) Key Name **always** contains at least one Part Separator (`/`).

But there is a problem.
We said a Key Name part is "an arbitrary sequence of non-zero bytes".

> This means "elektra/version" is a Key Name Part as well.

Author: What I wanted to say here is that "an arbitrary sequence of non-zero bytes" includes the string "elektra/version". That is why backslashes have to be escaped.

Then I gave an example of the escaping. I why the example is confusing, I should probably use a different one e.g. `/elektra\/version/info/back\\slash` with the parts `elektra/version`, `info` and `back\slash`.

For Unix paths, we could say that the canonical path is the shortest possible path that refers to a file.
In Elektra this doesn't quite work, but will use this definition for now.

> _Note:_ Only Escaped Key Names can be Canonical or Non-Canonical, so we normally omit the "Escaped" specifier.

Let's look at a few examples to get a feeling for Canonical and Non-Canonical Key Names.

| Non-Canonical           | Canonical          |
| ----------------------- | ------------------ |
| `/elektra/./version`    | `/elektra/version` |
| `/elektra/../version`   | `/version`         |
| `/elektra/.././version` | `/version`         |
| `/elektra///version`    | `/elektra/version` |
| `/elektra//../version`  | `/version`         |
| `/elektra/./../version` | `/elektra/version` |
| `/elektra/version/`     | `/elektra/version` |

So far this mostly follows, what we know from Unix paths.
However, the last two examples are somewhat different.

First, in Unix `/elektra/./../version` would be the same as `/version`, because `.` is just skipped and `..` removes `elektra`.
But in Elektra, we evaluate `..` first and only then skip `.`.

> `/elektra/./../version` is the same as `/elektra/version`, because `..` remove `.`.

Author: Yes, we already decided to match Unix behaviour. Specifically, `/./` is immediately discarded (reading parts left to right) and `/..` is allowed an maps to `/`.

- It contains a Namespace (i.e. the Namespace is not "Cascading"), but the Namespace Separator `:` is not followed by a `/`.
  (This mainly applies to Root Keys.)
- It contains a Namespace Separator `:`, but the substring before the first `:` is not one of: `meta`, `spec`, `proc`, `dir`, `user`, `system` and `default`.
- One of the Key Name Parts starts with an unescaped `@`.
  For example `/elektra/@version` is an invalid Escaped Key Name.
  Only `/elektra/\@version` would be valid.
- It contains an Illegal Escape Sequence (see below).
- It contains an Invalid Array Part (see below).
- Going from left to right one Key Name Part at a time, at any point, the number of Key Name Parts that are `..` is bigger than the number of other Key Name Parts.
  In Non-Canonical Key Names, the `.` Key Name Parts contribute to the number of other Key Name Parts.

An Escape Sequence is illegal, if (one of):

- It appears at the start of a Key Name Part and the escaped character is not one of: `/`, `\`, `.`, `#`, `%` or `@`.
- It appears anywhere else and the escaped character is not one of: `/` or `\`.

An Array Part is valid, if (one of):

- The whole Key Name Part is `#` or `#0`.

> - The Key Name Part is `#` followed by 1-19 digits, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^64 - 1`).

Author: Elektra uses a signed integer (`kdb_long_long_t`) for array indices. I think this is so that `-1` can be used as an error code for `elektraReadArrayNumber`. `ELEKTRA_MAX_ARRAY_SIZE` has not been updated yet, if you took that as a reference.

> - The Key Name Part is `#` followed by 1-19 digits, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^64 - 1`).

Reviewer 2: If I am honest, I did not look at the type of array indices. I just assumed that arrays indices use a 64-bit unsigned type, because of the text in the parentheses (`2^64 - 1`), which is `18446744073709551615` and not `9223372036854775807` (= `2^63 - 1`), according to my Python interpreter. Anyway, if we assume a signed 64-bit number, then I think we need to update the exponent:

```suggestion
- The Key Name Part is `#` followed by 1-19 digits, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^63 - 1`).
```

> - The Key Name Part is `#` followed by 1-19 digits, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^64 - 1`).

Author: You're right it should be 2^63 - 1.

# Key Names in Elektra

## Preface

This document is a full explanation of how Key Names work in Elektra.
In addition to this document, a reference Python implementation can be found in [keynames.py](keynames.py).
The goal this Python implementation is not to be fast, or to be used in any way other than as a reference.
If there are any discrepancies between this document, the Python implementation and the actual C implementation in [src/libs/elektra/keyname.c](../../src/libs/elektra/keyname.c), you should consider them as follows:

1. The C implementation is optimized for speed and much harder to maintain.
   It is mostly likely to be incorrect.
2. In most cases, this document outranks the Python implementation.
   There may, however, be cases where the language in this document was too vague and the Python implementation is actually correct.
3. If two of the sources agree, the third one is probably incorrect.
   Although again, if one of the agreeing sources is the C implementation it could still be the case that there is a mistake.

In any case: If you find a discrepancy, please file a bug report at https://issues.libelektra.org/new.

> > _Note:_ Mistakes happen.

Reviewer: The note can be removed. It is enough if people know what our prioritization is.

_To Elektra Developers:_ Feel free to add any unclear or previous incorrect examples to the test cases in [tests/ctest/test_keyname.c](../../tests/ctest/test_keyname.c).
These tests are very fast (1000+ test cases per second) and the more tests the better.

## 1. Key Name Parts and Namespaces

Before we get to Key Names proper, we need to talk about Key Name Parts and Namespaces.

Each Key is part of one of these _Namespaces_:

- Cascading
- Meta
- Spec
- Proc
- Dir
- User
- System
- Default

Each of these Namespaces has a very specific meaning, explained in [another section below](#12-namespaces-and-root-keys).

Apart from the Namespace, a Key Name is just a series of zero or more _Key Name Parts_.
Each Key Name Part is just an arbitrary (possibly empty) sequence of non-zero bytes.

So without knowing anything about how Key Names are written, we could say that there is a Key in the Namespace "System" with the Key Name Parts "elektra", "version" and "info".

> _Note:_ Not every such sequence, is a valid Key Name.
> For more information see [section 4](#4-valid-and-invalid-key-names)

### 1.1. Key Hierarchy

> You may have already seen elsewhere, that in Elektra Keys commonly look a bit like Unix paths:

Reviewer: ```suggestion
You may have already seen elsewhere, that in Elektra Keys commonly look like Unix paths:
```

```
/elektra/version/info
```

> _Note:_ How this representation works exactly and how namespaces come into play, will be explained in the next section.
> For now, we only care that there is some similarity to Unix paths.

This is no mistake.
Elektra's _Key Database (KDB)_ is designed to resemble a Unix filesystem.
In particular, the KDB has a similar hierarchy.
More generally, all Key Names exhibit this hierarchy.
By going back to thinking about a Key Name as a Namespace and a series of Key Name Parts, we can define this _Key Hierarchy_.

Each Namespace has a separate hierarchy.

> The relation between these, will be explored in the next section.

Reviewer: Overuse of forward references, give them all in the one note above (If you think it is needed. I think without any forward reference it is fine, too)

For now, we just look at a single Namespace.

In a Unix filesystems, we commonly talk about files and directories.
We also say a file is located within a directory.
But you might also know that in Unix "everything is a file".
This applies to directories as well, but "a file is located within a file" is a bit clunky, so you might say "file `A` is located below file `B`", if `B` is a file within the directory `A`.

> What makes `A` a directory is just the fact that there are other files below `A` [[1]](#footnote-1).

Reviewer: ```suggestion
What makes `A` a directory is just the fact that there can be other files below `A` [[1]](#footnote-1).
```

- A Key `Km` is _below_ `K`, if `Km` has `n+m` Key Name Parts and the first `n` Key Name Parts of `Km` are equal to the Key Name Parts of `K` (in the same order).
- A Key `K1` is _directly below_ `K`, if `K1` is below `K` and `K1` has `n+1` Key Name Parts.

Here are a few examples to show how this works in practice (using the Unix-path-like representation teased above):

| Key 1                   | Key 2                   | Relation                          |
| ----------------------- | ----------------------- | --------------------------------- |
| `/elektra/version/info` | `/elektra/version`      | "Key 1" is directly below "Key 2" |
| `/elektra/version/info` | `/elektra`              | "Key 1" is below "Key 2"          |
| `/elektra`              | `/elektra/version/info` | "Key 2" is below "Key 1"          |
| `/elektra/version/info` | `/elektra/version/info` | "Key 1" and "Key 2" are equal     |
| `/elektra/version/info` | `/elektra/data`         | no relation                       |
| `/elektra/data`         | `/elektra/version`      | "Key 1" and "Key 2" are siblings  |

You can think of the Key Hierarchy (within a single Namespace) as a big tree of Keys.

> Each node in the tree is a single Key `K` and the children of the node are the Keys that directly below `K`.

Reviewer: ```suggestion
Each node in the tree is a single key `K` and the children of the nodes are the keys that are directly below `K`.
```

>
> Instead, we recommend the terms _leaf Key_ and _non-leaf Key_, as these are commonly used for tree-like structures and their definitions fit perfectly.

#### 1.1.2. The "parent" confusion

As an inverse to "is below" we sometimes use "parent".
For example:

- `/elektra/version` is directly below `/elektra`, so `/elektra` is the (direct) parent of `/elektra/version`
- `/elektra/version/info` is below `/elektra`, so `/elektra` is a parent of `/elektra/version`
- `/elektra/version/info` and `/elektra/version` are below `/elektra`, so `/elektra` is the common parent.

This terminology can be easy to confuse, as "parent" is used for multiple different things.
If the context doesn't make it clear what "parent" means, you might consider more clear terms, e.g. other common terms for tree-like structures.
For example, you could use "ancestor" and "direct parent" as a clear differentiation.

There is also the term "(the) Parent Key".
In many cases, this is refers to a very specific Key in the given context that is a "common parent" to a certain set of Keys.
If you are working within such a context, be careful about "a parent Key" vs. "the Parent Key".

> 

Reviewer: Can you give an example here please?

### 1.2. Namespaces and Root Keys

We mentioned above that there are different Namespaces in Elektra.
Now we will explain their meaning.

To recap, Elektra knows these Namespaces:

- Cascading
- Meta
- Spec
- Proc
- Dir
- User
- System
- Default

We mentioned above that there are Key Names with zero Key Name Parts, i.e. just Namespace.
These are called _Root Keys_ (based on Unix's filesystem root, as well as the root of a tree).

> Lets explore them one by one:

Reviewer: Maybe separate "special" key names from others. Here is some overlap with doc/tutorials/namespaces.md

Author: Not sure what you mean by "special" key names... (I assume you are not referring to section 4.1 of this document)

Author: Your separation doesn't make sense. Cascading, Meta, Spec stand on their own and Proc, Dir, User, System, Default form a fallback chain. That's the only separation that would make sense IMO, but even that only makes sense **after** the Namespace Resolution section.

Reviewer: Yes, it is tricky to get a flow of a document right. For me "default" is special, as it is not an input but derived. So it depends on how exactly the other text around it is.

  The Key Hierarchy of the Namespace "Meta" is entirely separate from the Key Hierarchies in the other Namespaces.
  Without external information, you cannot determine from a Key Name with Namespace "Meta", which Key this Meta Key is attached to.
- Keys with Namespace **"Proc"** only exist in memory and are scoped to the current <b>Proc</b>ess.
- Keys with Namespace **"Dir"** are scoped to a single filesystem <b>Dir</b>ectory.
  They will (normally) be stored somewhere within this directory.
  Currently, there is no way of knowing which filesystem directory a Key with Namespace "Dir" is scoped to.
  Elektra only uses Keys scoped to the Current Working Directory.
- Similarly, Keys with Namespace **"User"** are scoped to a single **User**.
  They will (normally) be stored somewhere within the User's home directory.
  Again, there is no way of knowing which user a Key is scoped to and Elektra only uses Keys scoped to the current user, i.e. the one that executed the application.
- The Namespace **"System"** is what makes Elektra global.
  Keys with Namespace "System" are the same for all users of the system, independent of context.
  They are stored in system level directory.

> - Keys with Namespace "Default" are special.

Reviewer: ```suggestion
- Keys with Namespace **"Default"** are special.
```

  While you could create them manually, you normally don't want to.
  It is used for Keys with default values for Namespace Resolution (explained below).
  Keys with this Namespace are also in-memory only.
- Then there is **"Spec"**.
  This Namespace, like "Meta", is separate from the rest.
  We use it for Keys that are part of a specification used to describe other Keys.
  The Metadata of every Key with Namespace "Spec" describes a specification for the Keys that have the same Key Name Part but a different Namespace (except "Meta").
  So the Namespace "Spec" has a closer relation to the others than "Meta".
- Now just the Namespace **"Cascading"** remains.

>   This Namespace is used for Namespace Resolution (see below).

Reviewer: If you separate special namespaces from other namespaces, you can write something like this:

```suggestion
  This namespace indicates that the key can be from any namespace.
```

Author: The statement 
> The namespace "Cascading" indicates that the key can be from any (other) namespace

doesn't make sense. Even with the word "other", the key in question **is** of the namespace "Cascading". This is a namespace just like any other and can be used as such. The only difference is it's meaning. But every namespace has very precise meaning, so it doesn't make sense to differentiate on that basis either.

The statement 

> A cascading key names can refer to a key from any namespace. Such keys names always refer to a single key, but which namespace this key has depends on the context.

makes a little more sense. But it is still inaccurate. You can create a `Key` with a cascading key name and use it entirely separately from a `KeySet`. It would only be accurate, if the cascading key names could only be used for `ksLookupByName` and not for `keyNew` or `keySetName`. But that is of course not possible because of limitations in the implementation.

  It is the one that applications and end-users will use most commonly when interacting with Elektra.
  Keys with Namespace "Cascading" are never stored.

>   Not on disk and ideally also not in a `KeySet`.

Reviewer: ```suggestion
  Not on disk and normally also not in a `KeySet`.
```

There is also a certain ranking between the Namespaces "Proc", "Dir", "User", "System" and "Default".
Namely, that they override each other in exactly this order.
Given two Key Names with identical Key Name Parts, but one with Namespace "Dir" and one with Namespace "User", the one with Namespace "Dir" should be considered more specific and should be preferred.

> A special feature of Elektra is _Namespace Resolution_ (sometimes just Name Resolution, or Key Name Resolution).

Reviewer: Please use only one term and stick to it.

Author: I did stick to "Namespace Resolution" in this document, but other places (e.g. doc/tutorials/namespaces.md) talk about "resolving the (key) name". This want I wanted to indicate. I will use a different phrase to clarify that "resolving the name" == "resolving the namespace".

Author: But I can't look through all of the documentation to find every mention of "resolving a keyname" or something like that. I chose "Namespace Resolution", because it is more accurate to what I describe here. As footnote 2 indicates, the whole process used by `ksLookup` involves a bit more.

If you think about it this distinction makes sense. The Namespace Resolution is the process of finding the correct namespace. This is just a part of Name Resolution which finds the whole key name and therefore takes into account things like `override` metakeys.

Reviewer: Why not simply call it (cascading) lookup?

To not call it "resolving" is actually a good idea as "resolving" is associated with the file name resolver plugins.

Btw. there is a single occurrence of "resolved" (as verb!) in doc/tutorials/namespaces.md, so I would not call this an "established term".

Namespace Resolution is the process of finding an appropriate Namespace for a Key based on a Key Name with Namespace "Cascading".
It is most commonly used, when you to find which Key in the KDB should be used, based on a series of Key Name Parts.

To resolve the namespace, we just look at each of the Namespaces in the ranking defined above.
We then use the first Namespace where the Key actually exists.
Namespace Resolution is performed, when `ksLookup`/`ksLookupByName` is called with a Key Name with Namespace "Cascading" [[2]](#footnote-2).
<a id="ref-footnote-2"></a>
This is also done, if you call `kdb get` or `kdb set` with a Key Name with Namespace "Cascading".

## 2. Escaped Names

> The standard way to represent a Key Name in Elektra is this:

Reviewer: ```suggestion
We start with an example how to deconstruct following key name:
```

Author: I don't think it makes sense to say "we deconstruct a key name", when until this point the document defines a key name as "a namespace, plus a series of key name parts".

Maybe this would be better:

> The most common way to write down the key in the namespace "System" which has the key name parts `elektra`, `version` and `info` is this:

Then we have a transition from the definition above to the representation that we want to introduce.

```
system:/elektra/version/info
```

> This can be deconstructed into:

Reviewer: ```suggestion
```

For UNIX paths, we could say that the canonical path is the shortest possible path that refers to a file.
In Elektra this doesn't quite work, but will use this definition for now.

> _Note:_ Only Escaped Key Names can be Canonical or Non-Canonical, so we normally omit the "Escaped" specifier.

Let's look at a few examples to get a feeling for Canonical and Non-Canonical Key Names.

| Non-Canonical           | Canonical          |
| ----------------------- | ------------------ |
| `/elektra/./version`    | `/elektra/version` |
| `/elektra/../version`   | `/version`         |
| `/elektra/.././version` | `/version`         |
| `/elektra///version`    | `/elektra/version` |
| `/elektra//../version`  | `/version`         |
| `/elektra/./../version` | `/elektra/version` |
| `/elektra/version/`     | `/elektra/version` |

So far this mostly follows, what we know from UNIX paths.
However, the last two examples are somewhat different.

First, in UNIX `/elektra/./../version` would be the same as `/version`, because `.` is just skipped and `..` removes `elektra`.
But in Elektra, we evaluate `..` first and only then skip `.`.

> `/elektra/./../version` is the same as `/elektra/version`, because `..` remove `.`.

Author: @markus2330 @mpranj Do we want to keep that, or should we behave like UNIX paths? AFAIK what is described here was also the behaviour before the PR (I'd have to check to be sure).

Reviewer: Yes, sounds reasonable to change to the behavior towards UNIX paths.

Reviewer: Then we should also probably change that `/../../..` is a valid path (root stays root).

`/elektra/version/` and `/elektra/version` refer to the same Key and are always treated as the same Key Name.
The only exception are the Root Keys.
The Canonical Key Names for the Root Keys always end with a `/`.
In fact, we will see [later](#4-valid-and-invalid-key-names), that removing the `/` makes the Key Name invalid.

There also is a completely new addition in Elektra.
Elektra has a notion of _Array Parts_.
These are Key Name Parts that denote an array index.
How exactly these work, will be explored [later](#4-valid-and-invalid-key-names).
For now, we only need to know that they start with an `#` and their canonical form has `n` underscores followed by `n+1` digits.

A few examples for Array Parts:

| Non-Canonical    | Canonical           |
| ---------------- | ------------------- |
| `/elektra/#10`   | `/elektra/#_10`     |
| `/elektra/#1234` | `/elektra/#___1234` |

### 2.2. Other Escape Sequences

We already know, that `/` and `\` have to be escaped in an Escaped Key Name.
In addition two these to, there are a few more characters that have to be escaped.
However, these additional characters may **only** be escaped at the start of a Key Name Part.

> The characters in question are: `.`, `#`, `%`, `@`.

Author: Yes, but `/./` and `/../` still exist. If you want `keySetBaseName` to accept any string there has to be a way to escape these.

For example the Canonical Escaped Key Name for the Key with Namespace "Cascading" and Key Name Parts `elektra`, `.` and `version` is: `/elektra/\./version`.
The `\.` is the escaped form of `.`.
Using just `.` would make it a Non-Canonical Key Name, whose Canonical counterpart is `/elektra/version`, which does not have the wanted Key Name Part `.`.
Similarly, `#10` would make a Key Name Non-Canonical, but `\#10` won't.
Therefore, if you want a Canonical Key Name with a Key Name Part `#10` and not actually `#_10`,
must use e.g. `/elektra/\#10/version`.

<!-- TODO (kodebach): Unforunate escaping behaviour
        Unfortunately this means that both `/elektra/.version` and `/elektra/\.version` have the same unescaped form (`elektra`, `.version`).
        The same applies to `/elektra/%version` and `/elektra/\%version`.
        All of these are, however, Canonical Key Names.
        So the Key (CASCADING, `elektra`, `.version`) actually has two canonical forms....

> -->

Author: @markus2330 @mpranj As the comment says, I found this somewhat unfortunate behaviour. Not sure, what to do about it. We could say that `\.` is only a valid escape sequence, if the whole Key Name Part is `\.` or `\..`. I think that should probably be fine. For `\%` I'm not sure if that would work, because of the whole contextual value stuff that is so full of template code that I still don't understand it. I'm just very glad that it seems to still work apart from a tiny change related to the `default` namespace.

An alternative would be to canonicalize `\.version` and `\%version` into `.version` and `%version`, but again, not sure how the plays with the contextual value stuff. I think this will cause more problems, because AFAIK the contextual value stuff only uses the escaped name (and doesn't respect escaped slashes).

The best solution might be to ignore it for this PR and think of a fix before the 1.0 release.

Reviewer: I am okay with removing the special meaning of the `.` in the beginning of key name parts (the whole "hiding" feature is obsolete).

The `%` is not related to contextual values, it represents empty key name parts. You can forget about contextual values, there is no need to escape them, as they do not appear anymore in libelektra-core (they are resolved in the layer above).

Author: > The % is not related to contextual values, it represents empty key name parts.

Is there actually any need for that? It really makes very little sense to have an empty key name part anywhere. Even if you want to use a user supplied value as a key name part, you could just add a prefix/suffix to avoid empty key name parts.

> You can forget about contextual values, there is no need to escape them, as they do not appear anymore in libelektra-core (they are resolved in the layer above).

They may be resolved at the layer above, but it is still relevant to all of Elektra that such keys can have a special meaning. That was the whole point of #3223. We **need** to define a set of reserved characters that plugins, libraries and applications can use for special purposes (ideally in a semi-regulated fashion like `METADATA.ini` for metadata). Otherwise an application might use the key `/somewhere/%user%` and because the user mounted a certain plugin this turns into e.g. `/somewhere/kodebach` without the application knowing.

That is also why I am so insistent on having an easy to check a key name part for special meaning (e.g. checking just the first or first few bytes). If a component of Elektra needs to check the whole key name to make sure the key name won't be interpreted in a special way by some other component of Elektra, the whole thing will get insanely complicated and extremely slow.

Reviewer: > `%` Is there actually any need for that?

Yes, it is essential, otherwise keySetBaseName would not be able to work with any string. An empty string is as much string as any other. And yes, there are file formats that allow such things, e.g. TOML.

> you could just add a prefix/suffix

Of course you theoretically could always do millions of workarounds (practically it would not work: we would get millions of bugs in all storage plugins) but everybody still prefers if library functions like keySetBaseName just work without exceptions.

> They may be resolved at the layer above, but it is still relevant to all of Elektra that such keys can have a special meaning.

This is a wrong thought: only because something had a special meaning somewhere else, does not mean we need to do something about it. The % we get after the contextual evaluation do not have the contextual meaning anymore and it would be non-sense to give them a special treatment.

That `%` alone has a special meaning is just a coincident. (Like `.` and `.hidden` which is also simply a reuse of the same character.)

> because the user mounted a certain plugin this turns into

Contextual values unfortunately cannot be implemented as normal plugin as they happen at lookup.

> That is also why I am so insistent on having an easy to check a key name part for special meaning (e.g. checking just the first or first few bytes). If a component of Elektra needs to check the whole key name to make sure the key name won't be interpreted in a special way by some other component of Elektra, the whole thing will get insanely complicated and extremely slow.

We do not know if it would get slow. But yes, in general it fits better to Elektra if we have very short key part names which only have a single meaning (and a simple way to escape this meaning). The contextual values as implemented have been unnecessarily powerful but they are irrelevant in this escaping discussion anyway.

Author: > Yes, it is essential, otherwise keySetBaseName would not be able to work with any string. An empty string is as much string as any other.

That is **not** a reason IMHO. **We** define what `keySetBaseName` accepts and whether it makes sense that is should accept any string.

> And yes, there are file formats that allow such things, e.g. TOML.

This is slightly better argument but it is still beside the point.

Do you know of any instance where an application would use an empty key name part for its configuration where this cannot be easily avoided?

TOML, JSON, YAML and a million other formats might allow it. But if nobody actually uses it (because it is silly) it is still irrelevant to us.

> we would get millions of bugs in all storage plugins

AFAIK, there are just 2 plugins that can deal with any `KeySet` correctly: `quickdump` and (after #3507) `dump`. `mmapstorage` too, but only if absolute key names in the storage files are acceptable. Every other plugin runs into problems somwhere. Either the underlying storage format is limited or the plugin is buggy itself. In most cases, the limitation is that the storage file must be valid UTF-8 but Elektra's key names might not be.

> This is a wrong thought: only because something had a special meaning somewhere else, does not mean we need to do something about it.

The idea of Elektra is that an application is completely agnostic to the actual setup of the KDB. Therefore, if an application wants to use arbitrary key name parts (e.g. your example of WiFi SSIDs), then there has to be some way of guaranteeing that **no part** of Elektra no matter how obscure can interfere with this. Otherwise this misses the whole point of Elektra. The user should be in total control of their KDB and the application should just work.

> That `%` alone has a special meaning is just a coincident.

I know, and this entirely irrelevant.

> Contextual values unfortunately cannot be implemented as normal plugin as they happen at lookup.

Again, entirely irrelevant, as this was just an example.

Let's call it the `percent_user_percent` plugin. It's only job is replacing `%user%` with the current users name during `kdbGet` and `kdbSet`. This is a pretty stupid plugin, but there is no reason why this shouldn't be allowed.

Now someone else writes an application that does:
```c
Key * key = keyNew ("user/myapplication", KEY_END);
keySetBaseName (key, "%user%");
```

This is valid, because `keySetBaseName` accepts any string. The application then passes this key to `kdbSet`. The next time we start the application it calls `kdbGet` and does a `ksLookupByName (ks, "user/myapplication/%user%", 0)`.

Now we have a problem. If the user mounted `percent_user_percent`, the `ksLookupByName` won't work as expected.

I can give a million other examples like this. All of them just to show that we **need** some rules/restrictions for key names.

> We do not know if it would get slow.

Yes we do. It might be a minuscule difference for a single key name, but if for array parts alone there will be a noticeable difference at some point.

Currently this is the fastest way (that I know of) to check, if a single key name part `baseName` is a valid array part:

```c
int elektraArrayValidateBaseNameString (const char * baseName)
{
	const char * current = baseName;
	if (current == NULL || *current != '#')
	{
		return -1;
	}

	current++;

	if (*current == '\0')
	{
		return 0;
	}

	int underscores = 0;
	int digits = 0;

	while (*current == '_')
	{
		current++;
		underscores++;
	}

	while (isdigit ((unsigned char) *current))
	{
		current++;
		digits++;
	}

	bool underscoresCorrect = underscores == digits - 1;
	bool totalLengthCorrect = underscores + digits <= ELEKTRA_MAX_ARRAY_SIZE - 2;
	bool reachedEnd = *current == '\0' || *current == '/';

	if (underscoresCorrect && totalLengthCorrect && reachedEnd)
	{
		return underscores + 1;
	}

	return -1;
}
```

This assumes that `baseName` is an actual key name part that was validated by `elektraKeyNameValidate` at some point and not just an arbitrary string. For arbitrary strings you also need a `strncmp (&baseName[1 + underscores], "9223372036854775807", 19)`. It also only works for canonical key names.

Obviously `elektraKeyNameValidate` has to do all these checks, but it is really stupid that we need to do those checks again for already validated key name parts. If instead could just do `baseName[0] == '#'`, it would obviously be faster. How much faster might be unclear, but it is indisputable that a single byte comparison is faster than multiple loops and comparisons.

And this doesn't even take into account that because "`keySetBaseName` accepts any string" and the WiFi SSID example, you would technically have to check the escaped name as well to be entirely certain. Checking the escaped name is incredible complicated compared to even the above version of `elektraArrayValidateBaseNameString`, and definitely slower yet.

If you are unsure, why the escaped name has to be involved consider an application that uses keys of the form `/myapp/[WIFI-SSID]/autoconnecet` to store whether to automatically connect to a WiFi network. Since the WiFi-SSID is supplied by the user, it could be `#abc` or `#_12`. The application uses `keySetBaseName` to construct the keys, because it knows "`keySetBaseName` accepts any string". But how would `keySetBaseName` know that this time `#_12` should not be considered an array element when normally it should? It simply can't know, but let's assume you need to use some other function for array elements and `keySetBaseName` always escapes `#` (**not** the case right now). Even in that case, we will get the unescaped name parts `#abc` and `#_12`. To know that `#_12` is not supposed to be an array part, we would have to check the escaped name.

Alternatives:
1. (currently the case) `keySetBaseName` only escapes `#` when needed. In this case `#_12` will be treated as an array part and there is no way to detect that it shouldn't be.
2. There is no guarantee that `keySetBaseName` argument of `keySetBaseName` ends up being the last unescaped key name part. This breaks the expectation that `keySetBaseName (key, keyBaseName (key))` is a no-op on `key`, which could potentially cause **a lot** more trouble than restrictions on key names.
3. `keySetBaseName` doesn't accept any string. Not an option, because. Well, I don't know why, but you seem to be sure that it is not possible.

_**MOST OF ALL:**_ It might be still be fast enough to check the entire escaped and unescaped key name 10 times, but even if that were the case, it would still be slow, complicated and error-prone to write and maintain these checks. And because any plugin can currently claim special meaning for any key name part the whole exercise is irrelevant, because we don't even know what we have to check.

Reviewer: Thank you very much, I really appreciate that you try to get the best Elektra possible. As the discussion is scattered around everywhere and it gets impossible for others to understand what the outcome is, I started to write decisions in #3514. Please comment, as far as possible, there directly on the text so that we can come to an conclusion.

> We define what keySetBaseName accepts and whether it makes sense that is should accept any string.

Of course, I started the base_name.md decision to discuss this.

> Do you know of any instance where an application would use an empty key name part for its configuration where this cannot be easily avoided?

No and I would also not consider it as best practice to do something like this. KeySets, however, are not only for applications but also to represent contents of configuration files. And to have bug-free fully-conforming storage plugins is an important goal 86ceebb9daa7a841da543d7e62a685d207f9d220

> TOML, JSON, YAML and a million other formats might allow it. But if nobody actually uses it (because it is silly) it is still irrelevant to us.

If this would be an argumentation about some concrete configuration parameters, I would fully agree. But we discuss here the possibility if any application, which might not even exist at the moment, could ever need a key part name which is empty. I do not know the answer. Fact is, that we support it at the moment and that configuration file formats require its support, too. So it must be an extremely good argument to drop such a support.

> The idea of Elektra is that an application is completely agnostic to the actual setup of the KDB.

Yes, this was an idea/goal but it did not work. See capabilities.md

> then there has to be some way of guaranteeing that no part of Elektra no matter how obscure can interfere with this

Yes, I agree: characters.md

> but there is no reason why this shouldn't be allowed.

It would be allowed in the decision with `@`.

> All of them just to show that we need some rules/restrictions for key names.

Of course we need to escape `@` if we do not want its meaning. You are totally right about this and this is a big oversight on my part.

> elektraKeyNameValidate

does not even need to be called in a mmap scenario, so it does not matter if this is slow. Furthermore, I highly doubt that iterating several times over the key names will be noticeable compared to everything else going on as validation only happens at key construction. So these performance discussions are, as long as there is no benchmark showing a slowdown by a factor two of the whole `kdbGet` in a realistic scenario, simply irrelevant.

> #_12 should not be considered an array element when normally it should?

No, it is only an array element if the direct parent key has meta-data array, see array.md. So you implemented this already correctly afaik.

> This breaks the expectation that keySetBaseName (key, keyBaseName (key))

Yes, of course we want this. f8822fc4b47023a6732ae2f47175ee732a734755

> Well, I don't know why, but you seem to be sure that it is not possible.

:wink: It is my experience: we had many problems in storage plugins and after `keySetBaseName` started accepting everything, these problems disappeared.

> complicated and error-prone to write and maintain these checks.

Yes but much easier to maintain these checks at one place (and not duplicated for every storage plugin).

Author: > No, it is only an array element if the direct parent key has meta-data array, see array.md. So you implemented this already correctly afaik.

1. This is not possible for arrays in metadata. Metakeys cannot have metadata.
2. Relying on another key to decide the interpretation of a key is ridiculous, IMHO. That would be like saying this file is a JSON file unless the directory is a owned by root than it is a YAML file.

> It is my experience:  we had many problems in storage plugins and after `keySetBaseName` started accepting everything, these problems disappeared.

There will always be problems in storage plugins, especially since most of them are unmaintained. Just because you had bad experiences, doesn't mean it is impossible to provide an easy to use API for storage plugins in which `keySetBaseName` does not accept anything.

> Yes but much easier to maintain these checks at one place (and not duplicated for every storage plugin).

`storage` library...

Reviewer: >  This is not possible for arrays in metadata. Metakeys cannot have metadata.

I do not think that this is a problem as meta-arrays are pure convention and not a real thing (e.g. that are somehow serialized differently). I added this assumption in e533275e7f1bd90b5954051f536ba3b1301cb521

> Relying on another key to decide the interpretation of a key is ridiculous, IMHO. That would be like saying this file is a JSON file unless the directory is a owned by root than it is a YAML file.

Why do you think so? I think the other way: I would like to avoid having any semantics in the key names. For me semantics belong to metadata. Metadata is specifically designed to give keys semantics. For me the `#...` syntax is pure convention, it does not give any meaning and can be used anywhere where you want to count up something (also for non-arrays).

> There will always be problems in storage plugins, especially since most of them are unmaintained.

Which storage plugin is unmaintained? Imho only `ini`, which we will remove in #3491. We need to have realistic expectations of what a maintainer can do. In other free-software projects, like Linux, it is also common practice that big changes include the modification of all drivers, not that individual maintainers do it.

> Just because you had bad experiences, doesn't mean it is impossible to provide an easy to use API for storage plugins in which keySetBaseName does not accept anything.

I also did not claim that.

Author: > I do not think that this is a problem as meta-arrays are pure convention and not a real thing (e.g. that are somehow serialized differently).

That is just because, there is no storage plugin yet that supports metadata and also has a native array format.

Let's imagine `son`. It's like `ni` but for JSON. For this `ni` file:

```ini
my/key = 5

[my/key]
my/meta/#0 = a
my/meta/#1 = b
```

The `son` plugin could produce this JSON:

```json
{
  "keys": {
    "my/key": 5
  },
  "meta": {
    "my/key": {
      "my/meta": [ "a", "b" ]
    }
}
```

> I would like to avoid having any semantics in the key names. For me semantics belong to metadata. Metadata is specifically designed to give keys semantics. For me the `#...` syntax is pure convention, it does not give any meaning and can be used anywhere where you want to count up something (also for non-arrays).

This interpretation does make sense, yes. But then the metadata would have to be on the array elements. IMO the semantics of a `Key` should not change based on context.

Just think about a cascading key `/myarray/#0`. Where does the `array` metakey have to exist to make this an array element? Do you have to lookup `/myarray/#0` to get e.g. `user:/myarray/#0` and then `user:/myarray` has to have the metadata? Do we just do a lookup with `/myarray` and if `user:/myarray` doesn't exist, but `system:/myarray` has the `array` metadata its enough, and `user:/myarray/#0` is still an array element? What about the `spec` namespace?

The idea with the parent key is nice, but it is much more complicated and requires a lot more definitions than just "`#...` is an array element".

Reviewer: > there is no storage plugin yet that supports metadata and also has a native array format.

Yes, this is not covered by this decision. Obviously, you can simply guess that it is an array by the structure. This was what we were doing for normal arrays quite some time. This works quite well, except for empty arrays.

> But then the metadata would have to be on the array elements. IMO the semantics of a Key should not change based on context.

I agree it would be more beautiful if every key is standalone. Any suggestion? Adding `array/element` with a pointer to the parent to every element sounds very expensive (memory-wise), even if the meta-data is copied.

> Just think about a cascading key /myarray/#0. Where does the array metakey have to exist to make this an array element?

It does not work the reverse way. Where would this be needed?

> Do we just do a lookup with /myarray

Exactly, I added the description of that 1c729469fddb5125c0621ff369c87447b59aafb7

> and if user:/myarray doesn't exist, but system:/myarray has the array metadata its enough, and user:/myarray/#0 is still an array element?

I gave an example.

> What about the spec namespace?

What exactly do you mean? This is unaffected by the decision.

> The idea with the parent key is nice, but it is much more complicated and requires a lot more definitions than just "#... is an array element".

For most applications this will be true anyway. I do not think that many applications will abuse `#...` for anything else. The metadata `array` was mostly needed for empty arrays. I don't see any situation where it would be a problem to add the metadata (except of meta-arrays of course but this is a different topic).

Author: > Adding array/element with a pointer to the parent to every element sounds very expensive (memory-wise), even if the meta-data is copied.

I don't think `array/element` would need a value. Just `array/element = ` should be enough. Its obvious that the parent key is the array container. That way we could even reuse a single `Key` instance across the whole KDB.

> Re: array lookup

I commented on #3514.

> > The idea with the parent key is nice, but it is much more complicated and requires a lot more definitions than just "#... is an array element".
> 
> For most applications this will be true anyway. I do not think that many applications will abuse `#...` for anything else. The metadata `array` was mostly needed for empty arrays. I don't see any situation where it would be a problem to add the metadata (except of meta-arrays of course but this is a different topic).

You misunderstood. I didn't mean the application has to define more things. I meant it requires more definitions from us to specify how Elektra works.

Another somewhat confusing thing to me is: With #3447 the core definitely knows about array elements, since it adds the underscores when needed. But at the same time we kind of pretend the core doesn't know about array elements, because "arrays are just a convention".

Author: > I don't think `array/element` would need a value. Just `array/element = ` should be enough. Its obvious that the parent key is the array container. That way we could even reuse a single `Key` instance across the whole KDB.

Actually that doesn't work. We could have a structure like this:

```
myarray/#0/x = 1
myarray/#0/y = 2
```

Here we would have to add `myarray/#0` just to contain the `array/element =` metadata. That is not worth the additional memory. We would also still need a lookup to find out if `myarray/#0/x` is part of an array.

However, such a structure is also problematic with the current `array` metadata. You need to do a lot of work to find out, whether `#0` is actually an array element. I now remember that this is actually the reason why I originally wanted to make it so that `#` always indicates array elements.

Reviewer: Yes, I agree there is a lot of specification missing, also to clarify what an "array element" is. This huge PR, however, is not the right place for such a discussion. :wink:

>         So the Key (CASCADING, `elektra`, `.version`) actually has two canonical forms....

Author: This line is part of a TODO comment. Removing `keyIsInactive` and the special meaning of "starts with `.` " has not been implemented yet...

## 3. Unescaped Names

While the Escaped Name of a Key is nice for humans, it is not very well suited for machines.
The escaping of path separators makes it hard to find the Key Name Parts of a given Key Name.
The Escaped Name is also not well suited for sorting [[3]](#footnote-3).
<a id="ref-footnote-3"></a>

Both of these flaws are solved by the _Unescaped Name_.
In Unescaped Names we use a zero-byte as the Part Separator.
Since a Key Name Part cannot contain a zero-byte, we do not need an escaping mechanism for the path separator.

However, these zero-bytes mean that the Unescaped Name is not a printable string and therefore not human-readable.
This why we will only describe the Unescaped Name in the deconstructed form.

The Escaped Names from above correspond to the following Unescaped Names:

1. `system:/elektra/version/info`

>    - The byte representing the Namespace "System": `0x07`

Author: That would be confusing IMO. Then you would have to refer to the header file to find out what the enum entry actually means. It also doesn't have an advantage, to only use the enum entries. We can't easily change the values anyway, because that would impact the sort order of a KeySet.

But just in case, I will add a note stating that the byte values may have changed and that the `enum` in `kdb.h.in.c` is the source of truth

- The last byte is a zero-byte `0x00`.

For this reason Unescaped Names for Root Keys are 3 bytes long, with the last two bytes being zero-byte.
Between the second and the last byte, we find the Key Name Parts separated by a zero-byte `0x00`.

For Escaped Key Names it is easier to define, what makes an Invalid Key Name, than what makes a Valid Key Name, so we will go this route.
An Escaped Key Name is considered an Invalid Key Name, if any of the following are true:

- It is an empty string.
- The last character before the invisible null terminator is a backslash `\`.
  This called a "dangling escape", since we would expect another character after the escape character `\`.
- It contains a Namespace (i.e. the Namespace is not "Cascading"), but the Namespace Separator `:` is not followed by a `/`.
  (This mainly applies to Root Keys.)
- It contains a Namespace Separator `:`, but the substring before the first `:` is not one of: `meta`, `spec`, `proc`, `dir`, `user`, `system` and `default`.
- One of the Key Name Parts starts with an unescaped `@`.
  For example `/elektra/@version` is an invalid Escaped Key Name.
  Only `/elektra/\@version` would be valid.

> - It contains an Illegal Escape Sequence (see below).

Author: Originally "Illegal Escape Sequences" was supposed to be a whole sub-section. In that case the reference would have made more sense.

Otherwise, Array Parts (i.e. Key Name Parts starting with `#`) are invalid.

> _Note:_ The C-API does not allow you to construct a `Key` with an Invalid Key Name; for example `keyNew` (and `keyVNew`) will return `NULL`.

## 4.1. Special Key Names

Apart from Invalid Key Names, which cannot be constructed via the C-API, there are also _Reserved Key Names_.
These can be used with the C-API (`keyNew` returns a valid `Key *`), but there might be situations, in which `Key`s with such Key Names are treated differently.

A simple example is `/somewhere/_/below`.
If used with the globbing functionality, the Key Name Part `_` will be interpreted as "match anything that is not an array element".

If any of the following is true, a Key Name is a _Reserved Key Name_:

- A Key Name Part starts with `_`.
- A Key Name Part starts with `@`.
- A Key Name Part starts with `%` and the Key Name Part is more than one byte long.

<!-- TODO (kodebach): Is keyIsInactive still relevant? -->

There are also _Inactive Key Names_ (also _Hidden Key Names_).

> These are Key Names with a Key Name Part that starts with a `.`.

Author: @markus2330 @mpranj As the TODO says, is `keyIsInactive` still relevant or should it be removed any way?

Reviewer: Irrelevant, please remove.
